### PR TITLE
Upgrade codebase to use TypeScript and promises

### PR DESCRIPTION
**Pull request by ghost**
This is a major rewite of the code based on the original but now using TypeScript and ES6 promises. TypeScript `target` has been set to `ES3`, so it should work on most NodeJS installations as long as there is a Promise polyfill present.

This introduces some breaking changes, so if you think it is worthwhile to merge I think it is best to upgrade the package to version `2.0.0` or something similar.

**Note:** making Travis work required upgrading to node 4.0.0 I suspect because of typings/typings#722.

### CODE REVIEW

**Comment by justinmk** [comment]
ping @tarruda

File: .travis.yml [hunk 1]
```diff
@@ -3,7 +3,8 @@ language: node_js
 
 before_install:
   - eval "$(curl -Ss https://raw.githubusercontent.com/neovim/bot-ci/master/scripts/travis-setup.sh) nightly-x64"
+  - npm i -g typings
+  - typings install
```
Comment (tarruda) [inline .travis.yml:7]: AFAIK `typings` is not the standard way to install type declarations in typescript 2.0. I've been working on a private typescript project, and we use `npm install --save-dev @types/PACKAGE`. See [this](https://blogs.msdn.microsoft.com/typescript/2016/06/15/the-future-of-declaration-files/) for details.

If we're going to migrate to typescript, I prefer if we use the latest conventions.

File: Makefile [hunk 2]
```diff
@@ -2,12 +2,8 @@ TESTS = test/*.js
...
 
 test:
-	@# if any of the files contain 'debugger' statements, start with --debug-brk
-	@if find -name 'node_modules' -prune -o -type f -name '*.js' | xargs grep -q '^\s*debugger'; then \
-		./node_modules/.bin/mocha --debug-brk $(TESTS); \
-		else \
-		./node_modules/.bin/mocha $(TESTS); \
-		fi
+	./node_modules/typescript/bin/tsc
+	./node_modules/mocha/bin/mocha
```
Comment (tarruda) [inline Makefile:6]: Why did you remove the code to automatically add `--debug-brk`? It is actually an useful way to debug using mocha.

Also, the `tsc` command should probably go in a separate target:

```diff
diff --git a/Makefile b/Makefile
index 964cbaf..ea298e0 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,9 @@
 TESTS = test/*.js

+tsc:
+       @./node_modules/.bin/tsc

-test:
+test: tsc
        @# if any of the files contain 'debugger' statements, start with --debug-brk
        @if find -name 'node_modules' -prune -o -type f -name '*.js' | xargs grep -q '^\s*debugger'; then \
                ./node_modules/.bin/mocha --debug-brk $(TESTS); \
@@ -14,4 +16,4 @@ watch:
        @./node_modules/.bin/nodemon -e js --exec 'make test'


-.PHONY: test watch
+.PHONY: tsc test watch
```

File: index.js [hunk 3]
```diff
@@ -1,107 +1,102 @@
...
+                this.emit('notification', event_name.toString(), args);
+                break;
+            }
+            default:
+                this._encoder.write([1, 0, 'Invalid message type', null]);
+        }
+    };
+    return Session;
+}(events_1.EventEmitter));
 module.exports = Session;
```
Comment (tarruda) [inline index.js:102]: Please don't commit compiled files. Instead, just change the `prepublish` hook to invoke tsc.

File: tsconfig.json [hunk 4]
```diff
@@ -0,0 +1,9 @@
+{
+  "compilerOptions": {
+    "target": "es3",
```
Comment (tarruda) [inline tsconfig.json:3]: Why ES3? Isn't ES5 supported in node.js since its very early versions?

**Review by tarruda — CHANGES_REQUESTED** [review]
This still needs some serious work, and to be honest, I'm not sure it is worth the trouble.

First, I don't see much benefit of rewriting in typescript unless you make heavy use of type information, and you have hardly used any. For example, there's a type declaration for msgpack5 and you haven't imported or installed any.

Second, the only change for consumers of this library is that `session.request` returns a promise instead of accepting a callback, but this could be very easily achieved with a call to `Bluebird.promisify`. We could even provide a promisified version of the `request` method without going through all this trouble.

Honestly, I suggest that you simply create a separate package that provides a promisified version of the library that depends on this(it would be much easier to inherit from the existing Session class and override `request` to return a promise).

If you really want this to be merged, there must be a lot more benefit. What I'm seeing hardly justifies breaking compatibility, much less a full rewrite.

**Comment by ghost** [comment]
Thanks for the feedback, I get what you are saying. I'll be looking into it if I find some time. In the meantime, I'll close it.
